### PR TITLE
[WiP] Enable FindBugs for the project

### PR DESCRIPTION
**Pull request by oleg-nenashev**

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@reviewbybees

**Comment by christ66** [comment]
:bee:

File: src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ public synchronized String format(LogRecord record) {
                 record.getThrown().printStackTrace(out);
                 out.close();
                 buf.append(writer.toString());
-            } catch (Exception e) {
+            } catch (Throwable e) {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java]: Needs error processing

File: src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java [hunk 2]
```diff
@@ -68,7 +68,9 @@ public synchronized String format(LogRecord record) {
                 record.getThrown().printStackTrace(out);
                 out.close();
                 buf.append(writer.toString());
-            } catch (Exception e) {
+            } catch (Error error) {
+                throw error; // We propagate all Errors upstairs
+            } catch (Throwable e) {
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java]: Uh, what is the purpose of this change exactly?

File: src/main/java/com/cloudbees/jenkins/support/util/ArrayUtil.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.jenkins.support.util;
+
+//TODO: Get rid of it when the method becomes available in Jenkins core (hudson.Uti)
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/util/ArrayUtil.java]: Is that some PR? Which?

File: src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java [hunk 4]
```diff
@@ -411,8 +423,20 @@ private void addMasterJulLogRecords(Container result) {
...
+        
+        final File[] logFiles = jenkins.getRootDir().listFiles(ROTATED_LOGFILE_FILTER);
+        if (logFiles == null) {
+            LOGGER.log(Level.WARNING, "Cannot add master logs to the bundle. Cannot access log files");
+            return;
+        }
+        
         // log records written to the disk
-        for (File file : SupportPlugin.getRootDirectory().listFiles(new LogFilenameFilter())){
+        for (File file : logFiles){
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java]: :bug: this appears to be changing the implementation to something different and incorrect for no apparent reason. I think you want

```java
for (File file : ArrayUtil.fromList(SupportPlugin.getRootDirectory().listFiles(new LogFilenameFilter())) {
```

Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java]: (copy-paste error I suppose)

Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java]: Thanks for catching it!

File: src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java]: The idea is to catch all kinds of throwables excepting ones known to be an "unrecoverable" issue.

File: src/main/java/com/cloudbees/jenkins/support/util/ArrayUtil.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/util/ArrayUtil.java]: hudson.Util is not a right place. I'll just get rid of this method, because there are only 2 usages in the code

File: src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java [hunk 5]
```diff
@@ -238,14 +239,17 @@ public GetSlaveDigest(FilePath rootPath) {
...
             final File rootPath = new File(this.rootPathName);
-            for (File file : rootPath.listFiles()) {
-                if (file.isFile()) {
-                    try {
-                        result.append(Util.getDigestOf(new FileInputStream(file)))
-                                .append("  ")
-                                .append(file.getName()).append('\n');
-                    } catch (IOException e) {
-                        // ignore
+            final File[] files = rootPath.listFiles();
```
Comment (Vlatombe) [inline src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java]: Use FileUtils.listFiles here as well?

Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java]: Yes, makes sense. I've missed this call

File: src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java [hunk 6]
```diff
@@ -365,11 +365,23 @@ private void addLogRecorders(Container result) {
...
     private void addOtherMasterLogs(Container result) {
-        for (File f : Jenkins.getInstance().getRootDir().listFiles(ROTATED_LOGFILE_FILTER)) {
+        final Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
+            LOGGER.log(Level.WARNING, "Cannot add master logs to the bundle. Jenkins instance is not ready");
+            return;
+        }
+        
+        final File[] logFiles = jenkins.getRootDir().listFiles(ROTATED_LOGFILE_FILTER);
+        if (logFiles == null) {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java]: And another one...

File: src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/support/SupportLogFormatter.java]: According to the discussion in another thread, such pattern doe not make much sense. I'll cleanup it

**Comment by oleg-nenashev** [comment]
@Vlatombe WDYT about the current version?

**Comment by Vlatombe** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
@reviewbybees done

Squashing commits

**Comment by oleg-nenashev** [comment]
Jenkins.getInstance() also needs some love, but let's keep it for future iterations

**Comment by oleg-nenashev** [comment]
Tests are also unstable...

File: src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java [hunk 7]
```diff
@@ -365,11 +366,17 @@ private void addLogRecorders(Container result) {
...
      */
     private void addOtherMasterLogs(Container result) {
-        for (File f : Jenkins.getInstance().getRootDir().listFiles(ROTATED_LOGFILE_FILTER)) {
+        final Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
+            LOGGER.log(Level.WARNING, "Cannot add master logs to the bundle. Jenkins instance is not ready");
+            return;
+        }
+        
+        for (File f : FileUtils.listFiles(jenkins.getRootDir(), null, false)) {
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/JenkinsLogs.java]: this no longer uses the filter (ROTATED_LOGFILE_FILTER)

File: src/main/java/com/cloudbees/jenkins/support/impl/MetricsContent.java [hunk 8]
```diff
@@ -7,6 +7,7 @@
 import com.codahale.metrics.json.MetricsModule;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializationFeature;
+import hudson.model.Computer;
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/MetricsContent.java]: seems to be unwaranted

**Comment by jtnord** [comment]
:bug: on the missing filter

**Comment by oleg-nenashev** [comment]
Urgh... I've thought I've already fixed it
